### DISCUSSION TITLE
when migrating to v5 i have problem with pageParam inside my function. Problem TS2339: Property 'pageParam' does not exist on type '{ queryKey: QueryKey; signal: AbortSignal; meta: Record<string, unknown> | undefined; }'.

### DISCUSSION THREAD

**Discussion by ghblacktrade** [TanStack/query#6370] [General]
`export const getNotificationControllerGetNotificationInfiniteQueryOptions = <
    TData = Awaited<ReturnType<typeof notificationControllerGetNotification>>,
    TError = AxiosError<NotificationListEntity[]>,
>(
    params?: NotificationControllerGetNotificationParams,
    options?: {
        query?: UseInfiniteQueryOptions<
            Awaited<ReturnType<typeof notificationControllerGetNotification>>,
            TError,
            TData
        >;
        axios?: AxiosRequestConfig;
    },
) => {
    const {query: queryOptions, axios: axiosOptions} = options ?? {};

    const queryKey = queryOptions?.queryKey ?? getNotificationControllerGetNotificationQueryKey(params);

    const queryFn: QueryFunction<Awaited<ReturnType<typeof notificationControllerGetNotification>>> = ({pageParam}) =>
        notificationControllerGetNotification({cursor: pageParam, ...params}, axiosOptions);

    return {queryKey, queryFn, ...queryOptions} as UseInfiniteQueryOptions<
        Awaited<ReturnType<typeof notificationControllerGetNotification>>,
        TError,
        TData
    > & {queryKey: QueryKey};
};`

**Comment by TkDodo** [comment]
show a typescript playground please.

https://www.totaltypescript.com/need-a-playground

**Reply by ghblacktrade** [reply]
https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgOQhgXQW02IiQBkEBnGAURQRgE84BvAKDj7mxRgFYABMAgjABccBlGQBzANy9+wFuwCSogPwy5iuAB84SAK4AbCyv5x1xdgBU2YYDMLFS5SkmYO2zq42-Aii+jDySMqqfAC2wHR0mApushGKwXxCmGFwAEYQEBbCSJlwDsXhkdG26QopUO5EJGQUCFSO8vVoZeyuTZ6tPoHAKgC+3JOgkLACVAymzV5tVADCVBFFxVAA4sAwHi3e7UhwALxwABQxcGCYUJixdHoES0Mn6yhQW2h7B2-HKgABXujzoABobhAwD5njJxCB2nQAErAACOZgSME+pAUkIAlDIgd9YvRgAAeBFI1F0SBIOgUw7LHy0Bh+VhsADaAF0AHy884Cni2IQwMxQU6YREQOgAOlEwCwlhgspSMEuAAMAPSYMAILVIAEreka8FcG62WVW6GwyG2Wx3B5PGScK2yx1gs1um0nZ7u0FPMZ2-hjfHjFSTbjTaDwbDzeBqpnvNYbb5WX77JOApCaJCkJCsYAARUxUDYAHkYb7znByTdHAARCiYGviADumELonJqLFEpG5L6wAg6EWg2zn026d2maNPn5wb4jkYUG+UFb0roK7X5KzxtZTA0XL5kN51wdAbhr3Hxsnae2fz3PhBTohUKr8xewvtfAxaDYLwAKoMrm+aFiW-6VrCdY-j+7adqg3a9uKSADkOI5jkct6pj8M7-De868ousFLtu0DESRjbNha-C8mUthSkiLxUjKqJ-gwOIIDUIYEoK5q2HG9LwJwf5ljIokVh+9JmoxMoyLJdBQb6Yw1j68xwDoOhcGMEYCfGcASQA0sAHAXBJSmfrKRkmRpWmJnOHw4dOj4OVQEFlsZbCXB6TxhpMelCQZpZsAAYkgMjuaFZhINgPiUh2XY9vsKFoS4w6joaBGOV8uEuVlVD8gKFyXJwdwpC+jyhoKNF8JlWE+HeuWzvlSAldg4p0NAMhlcAFWxF6Vo+XQQZwApFn0n5NyiihXDWWwZoSWFA1WcF43DaNdBwMBwCgcg4GrVJdAwbB8GJch-ZpYOaUYXVzLZVOD7NfVJwLjVcDLqu5FvVRMCYDcAoAGSzcFnkRSDJk6dwkNAA

**Reply by TkDodo** [reply]
sorry but this is not helpful, you need to import types from libs that you are using and you need to either define or stub out your own types. You just dumped the code snippet into ts playground and now it's all red. If you want help, please do the necessary work to make it easy for maintainers to help you. right now this is just a waste of my time, sorry

**Reply by ghblacktrade** [reply]
oh, sorry for my hurry 

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwK4GcCmBJAdgM2F4GNARSTSgE8AaOVNAWSRgEMZgIsa7TyKBfODigQQcAOQABFlhQsAxgGsA9FDTM5MALQBHMpTEBuAFChIsODApg0iI3HtwelAGJIsG9pzsOnFANJo1N72AKro2HgERL4A8mBsHChURgJCIuJSzDLyyqrqWrq8hibg0PAAVHDMKFUAHuw1aaJizPUQKMWmZRZWNggAgm0oAKJQwlA0gw0ASmiFsgDCHHgA5pNDsyiQMmipws2tDcVGaLVm8JbWcAByEGx4cqyeACq9cAC8cAAUl2gQODc7sAHk8OK9rABKADaCkC-x61nht3uwEeCSw4LQAF1jEYlEo4GgUAAbfBaAAmwBQzAARsS0JosKctKSmXApL0UHIoMB4poiayYEosBBNKpyWg5MTmKoTmdunJEvBkcDUaCMW9PghgnB8DgIAAucR6iBiZIOOAAdxlBCwKyNYmtUFtKzNOpwzGAxIdHq9boE1TgiuyuNO5wRNhVIPRwywbEsmI+31+SKB0c8sfjFEx0NhFHhKYBUbVMbjhGzvRxRjxBIFZM0lOpdIZTNqLPwNg51i5PL5dbjwtF4sl0tlYYVSsBKLRGbLCc1tgtKAosjQIAdy9XIDdFroUAde53DmDMGExIA+mA5DSHSez5fr0f7ESUA6X-6qjVg7JcXLw78p1VGcwR5FYVnIJMfl6VNp3VZ5QPAqBczhAFC0A9MQOAMDyCrGtCRJetG1pelGWZTRWU7X4e15LR+yFEUxTQCUpRlNA-wnbJ0JLF4EIgrUdVUFYqSIfdxEE4TyCfIMAAssnA88109b1xDkWS7TQBSQCUqTVFkaANLAaoUEtaByQdXSYH0y8jJMqBySklAkDAc4Nyc84pL3c8JXpIgD3QKAvLQHy0Cku8IAvK8aXPIh6RANBTwoW8OFPcKHyimK13i-RkgDL8lVDeVzAA2ZmHJRNPigxEARKsremQ-NUOg6q1Fq6xcPxfDBQbKliJbMiKPZKjuRo-kCIHBjhxYsdCvgb94Bq8rFwcZhiWUlpVqkplLQdLaP0DOaCv-N4ABkIEeelILQ07zrQeqCyauBrpW7FcQ6ujuqbEjW3bNkuyJYa+zG+ihyYkdWPY8w5ses7nqTbULTQLA3ywHSkHMpA9rykNq2rcdzDJcgPTkSM024jhjuEzNyyW+xuTUIhyX6GAjVkHk7WMBG5woDByQAfhZ098BWOAAB84CwJBVo5hxEazTEjWLYCsCp+drGl+xgDMuBWaF9W4DilBqXAgW2ZWPW8i1mkIHCtQsD1+N6RN3WdUF7DRMVuDeKge3egV0mlcxYw+FxmaIy4pWljjM96SgABxeKPfRAAFGVmBAGp+ItOQkCgFBoH57XBfZnUUGAAAvNAC4lkAaXIc2WoLhbeh96wC8Tl5m5LlhYAAEVYSuneLznyT7ogC51oeZa5nnx6Ls2dWJGH6QLp76SD388dmycRVg9FI5S1byHjmB244SCdUMqA05QNv-fVffo6PhO7+T1P0-NBwIHiTwb6NKZ2lmPMGAkdVjJAhEaJO+wqRoAADz-xQJsbY6AYGnywBTWQKsKBQixAAPhwR8fB8MHCqBgDnLAdQGgADoJQekljASh4EYBfAAAZKGYGAYAg5d4-2YTQIhFp7CUKEV-dESQdQWkvtfI0CAhGUMke-OAsiRE-15nIt+KA+Af3sHwCE68Q7hihow1BD9wox2Pqg3wAQKDnwkeo2+3COAmMPnHZ+DisApyvu-IwEICE0zgCQshcAoSsPYZwneQF1QoF4YooRXx5E1F5kE+JWI4BGmwRCFJ+18opEOhxWQcAjEv08E4sxriInogiPgQgJA9AUDiKIpMMCdTPFHswJM-RrTVPJDA2YpDnSYhgWhcJGEsAlKfifIpHA8FaLgM8UY4x2lDHmdAFBky0GUy5tgnByQcFfAvnYv2bixkuImW4jx18ZnKMSAXfhAjCiUALmETAuAqnRFqfUn+TSBHfLgB0z0DMenxTIQMoZazjnmLWdM8RPy5ljGgDMn5szWnQvsDgvWFpDjtALvAwBZBFjLCwnrTR3jfG3Khgge5iU4CUo+YkGgmLXwUPaLSmQAhPhXJkHAXmiSEB8FxFnSclKrFJhpd-a5lChWBC5YkwpRzkqPxORY2pVi4nqN0dWAVnFKXOCRo4Wprh3DojgZ0gFvTgW9EGQ9YZZNRnytMeM1BeD8EVQQIZcC5yQA6IISi8WYK7XOIhW4r4CBs652gEaN1aAPU0FkfEzRTKUAspQOqgSQLnSIEldQal+rOAxIle8sVrLPxwCeZUqINTeBJq+T8v5XTAV9I1NYS1VVfVyqjvaxVkLtk+theMBFDgWmsGYDqfBAAyDNyrAhGksYEPlOSjBAA

**Reply by TkDodo** [reply]
the type `QueryFunction` has 3 generics and you're not passing them:

https://github.com/TanStack/query/blob/7c87c3d7b7fb0edf3a9ac69e2f29c0a3be07a669/packages/query-core/src/types.ts#L31-L35

Overall, it looks like what you're trying to do is overly complex. Have a look at the new `infiniteQueryOptions` helper funtion
